crates/ruff_cli/src/commands/format.rs

```diff
@@ -145,24 +146,111 @@ fn format_path(
+    source_type: PySourceType,
+    path: Option<&Path>,
+    settings: &FormatterSettings,
+) -> Result<Option<SourceKind>, FormatCommandError> {
+    match source_kind {
+        SourceKind::Python(unformatted) => {
+            let options = settings.to_format_options(source_type, unformatted);
+
+            let formatted = format_module_source(unformatted, options)
+                .map_err(|err| FormatCommandError::Format(path.map(Path::to_path_buf), err))?;
+
+            let formatted = formatted.into_code();
+            if formatted.len() == unformatted.len() && formatted == *unformatted {
+                Ok(None)
+            } else {
+                Ok(Some(SourceKind::Python(formatted)))
+            }
+        }
+        SourceKind::IpyNotebook(notebook) => {
+            if !notebook.is_python_notebook() {
+                return Ok(None);
+            }
+
+            let mut output = String::with_capacity(notebook.source_code().len());
+            let mut source_map = SourceMap::default();
+            let mut last: Option<TextSize> = None;
+
+            // Format each cell individually.
+            for (start, end) in notebook.cell_offsets().iter().tuple_windows::<(_, _)>() {
+                let range = TextRange::new(*start, *end);
+                let unformatted = &notebook.source_code()[range];
+
+                let options = settings.to_format_options(source_type, unformatted);
+
+                // Format the cell.
+                let formatted = format_module_source(unformatted, options)
+                    .map_err(|err| FormatCommandError::Format(path.map(Path::to_path_buf), err))?;
+
+                // If the cell is unchanged, skip it.
+                let formatted = formatted.as_code();
+                if formatted.len() == unformatted.len() && formatted == unformatted {
+                    continue;
+                }
+
+                // Add all contents from `last` to the current cell.
+                let slice = &notebook.source_code()
+                    [TextRange::new(last.unwrap_or_default(), range.start())];
+                output.push_str(slice);
+
+                // Add the start source marker for the cell.
+                source_map.push_marker(*start, output.text_len());
+
+                // Add the cell itself.
+                output.push_str(formatted);
+
+                // Add the end source marker for the added cell.
+                source_map.push_marker(*end, output.text_len());
+
+                // Track that the cell was formatted.
+                last = Some(*end);
+            }
+
+            // If the file was unchanged, return `None`.
+            let Some(last) = last else {
+                return Ok(None);
+            };
+
+            // Add the remaining content.
+            let slice = &notebook.source_code()[usize::from(last)..];
+            output.push_str(slice);
+
+            // Update the notebook.
+            let mut notebook = notebook.clone();
```

Comment:
Could we call-by-value with `SourceKind` instead of call-by-reference or does a caller need this for error handling?

Reply:
In the `stdin` case, we still need access to the unmodified source so it can't be moved. But I can make this clearer with a custom enum.